readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,340 @@
+            order_by='path',
+            sort='asc',
+            membership=True,
+        )
+
+        try:
+            for repo in repos:
+                self.create_repository(repo)
+        except (TypeError, ValueError):
+            log.exception('Error syncing GitLab repositories')
+            raise Exception(
+                'Could not sync your GitLab repositories, try reconnecting '
+                'your account')
+
+    def sync_organizations(self):
+        orgs = self.paginate(
+            '{url}/api/v4/groups'.format(url=self.adapter.provider_base_url),
+            per_page=100,
+            all_available=False,
+            order_by='path',
+            sort='asc',
+        )
+
+        try:
+            for org in orgs:
+                org_obj = self.create_organization(org)
+                org_repos = self.paginate(
+                    '{url}/api/v4/groups/{id}/projects'.format(
+                        url=self.adapter.provider_base_url,
+                        id=org['id'],
+                    ),
+                    per_page=100,
+                    archived=False,
+                    order_by='path',
+                    sort='asc',
+                )
+                for repo in org_repos:
+                    self.create_repository(repo, organization=org_obj)
+        except (TypeError, ValueError):
+            log.exception('Error syncing GitLab organizations')
+            raise Exception(
+                'Could not sync your GitLab organization, try reconnecting '
+                'your account')
+
+    def is_owned_by(self, owner_id):
+        return self.account.extra_data['id'] == owner_id
+
+    def create_repository(
+            self, fields, privacy=DEFAULT_PRIVACY_LEVEL, organization=None):
+        """
+        Update or create a repository from GitLab API response.
+
+        :param fields: dictionary of response data from API
+        :param privacy: privacy level to support
+        :param organization: remote organization to associate with
+        :type organization: RemoteOrganization
+        :rtype: RemoteRepository
+        """
+        repo_is_public = fields['visibility'] == 'public'
+        if privacy == 'private' or (repo_is_public and privacy == 'public'):
+            try:
+                repo = RemoteRepository.objects.get(
+                    full_name=fields['name_with_namespace'],
+                    users=self.user,
+                    account=self.account,
+                )
+            except RemoteRepository.DoesNotExist:
+                repo = RemoteRepository.objects.create(
+                    full_name=fields['name_with_namespace'],
+                    account=self.account,
+                )
+                repo.users.add(self.user)
+
+            if repo.organization and repo.organization != organization:
+                log.debug(
+                    'Not importing %s because mismatched orgs',
+                    fields['name'],
+                )
```

Comment:
Wonder if this should be a higher priority? Will this cause breakage for the user?

Reply:
Honestly. I'm not sure. Github and Bitbucket code do exactly the same.

It's kind of a weird case, because it should happend that user `U` created the repo `R` with organization `O`. Then, user `UU` tries to import the _same_ repo `R` but with organization `OO` and that generates this error, right?

Reply:
That is correct. Probably no need to surface this any harder, it's a weird edge case.